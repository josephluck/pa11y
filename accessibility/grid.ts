import { ContrastResult, contrast } from "./contrast";
import { split } from "../utils/pairs";

type Grid = GridRow[];

type GridRow = GridItem[];

type GridItem = ContrastResult & {
  fg: string;
  bg: string;
};

export const grid = <T extends string>(colors: T[]): Grid =>
  split(colors, colors).map((row) =>
    row.map((cell) => ({
      fg: cell[0],
      bg: cell[1],
      ...contrast(cell[0], cell[1]),
    }))
  );
