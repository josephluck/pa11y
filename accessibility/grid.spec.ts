import { grid } from "./grid";

describe("Grid", () => {
  it.each([
    { colors: ["#FFFFFF", "#000000"] },
    { colors: ["#FFFFFF", "#111111", "#000000"] },
    { colors: ["#FFFFFF", "#AAAAAA", "#444444", "#000000"] },
    { colors: ["#FFFFFF", "#AAAAAA", "#777777", "#444444", "#000000"] },
  ])("generates a grid of accessibility results", (row) => {
    expect(grid(row.colors)).toMatchSnapshot();
  });
});
