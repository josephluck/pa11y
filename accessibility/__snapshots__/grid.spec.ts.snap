// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid generates a grid of accessibility results 1`] = `
Array [
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 1,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 21,
      "fg": "#FFFFFF",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 21,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 1,
      "fg": "#000000",
    },
  ],
]
`;

exports[`Grid generates a grid of accessibility results 2`] = `
Array [
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 1,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#111111",
      "contrast": 18.883060964594996,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 21,
      "fg": "#FFFFFF",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 18.883060964594996,
      "fg": "#111111",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#111111",
      "contrast": 1,
      "fg": "#111111",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 1.1121078324840545,
      "fg": "#111111",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 21,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#111111",
      "contrast": 1.1121078324840545,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 1,
      "fg": "#000000",
    },
  ],
]
`;

exports[`Grid generates a grid of accessibility results 3`] = `
Array [
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 1,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#AAAAAA",
      "contrast": 2.3231230535045992,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#444444",
      "contrast": 9.739769120526205,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 21,
      "fg": "#FFFFFF",
    },
  ],
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 2.3231230535045992,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#AAAAAA",
      "contrast": 1,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#444444",
      "contrast": 4.192532593498678,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 9.039555596643915,
      "fg": "#AAAAAA",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 9.739769120526205,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#AAAAAA",
      "contrast": 4.192532593498678,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#444444",
      "contrast": 1,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 2.156108603821344,
      "fg": "#444444",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 21,
      "fg": "#000000",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#AAAAAA",
      "contrast": 9.039555596643915,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#444444",
      "contrast": 2.156108603821344,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 1,
      "fg": "#000000",
    },
  ],
]
`;

exports[`Grid generates a grid of accessibility results 4`] = `
Array [
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 1,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#AAAAAA",
      "contrast": 2.3231230535045992,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#777777",
      "contrast": 4.478089453577214,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#444444",
      "contrast": 9.739769120526205,
      "fg": "#FFFFFF",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 21,
      "fg": "#FFFFFF",
    },
  ],
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#FFFFFF",
      "contrast": 2.3231230535045992,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#AAAAAA",
      "contrast": 1,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#777777",
      "contrast": 1.9276161229693327,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#444444",
      "contrast": 4.192532593498678,
      "fg": "#AAAAAA",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#000000",
      "contrast": 9.039555596643915,
      "fg": "#AAAAAA",
    },
  ],
  Array [
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#FFFFFF",
      "contrast": 4.478089453577214,
      "fg": "#777777",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#AAAAAA",
      "contrast": 1.9276161229693327,
      "fg": "#777777",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#777777",
      "contrast": 1,
      "fg": "#777777",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#444444",
      "contrast": 2.174983153305664,
      "fg": "#777777",
    },
    Object {
      "AA": true,
      "AAA": false,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
      ],
      "bg": "#000000",
      "contrast": 4.68949989000882,
      "fg": "#777777",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 9.739769120526205,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
      ],
      "bg": "#AAAAAA",
      "contrast": 4.192532593498678,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#777777",
      "contrast": 2.174983153305664,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#444444",
      "contrast": 1,
      "fg": "#444444",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 2.156108603821344,
      "fg": "#444444",
    },
  ],
  Array [
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#FFFFFF",
      "contrast": 21,
      "fg": "#000000",
    },
    Object {
      "AA": true,
      "AAA": true,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
        "AAA",
      ],
      "bg": "#AAAAAA",
      "contrast": 9.039555596643915,
      "fg": "#000000",
    },
    Object {
      "AA": true,
      "AAA": false,
      "AAALarge": true,
      "AALarge": true,
      "accessibilityLevels": Array [
        "AALarge",
        "AA",
        "AAALarge",
      ],
      "bg": "#777777",
      "contrast": 4.68949989000882,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#444444",
      "contrast": 2.156108603821344,
      "fg": "#000000",
    },
    Object {
      "AA": false,
      "AAA": false,
      "AAALarge": false,
      "AALarge": false,
      "accessibilityLevels": Array [],
      "bg": "#000000",
      "contrast": 1,
      "fg": "#000000",
    },
  ],
]
`;
